datafusion/optimizer/src/eliminate_limit.rs

```diff
@@ -63,17 +63,23 @@ impl OptimizerRule for EliminateLimit {
     > {
         match plan {
             LogicalPlan::Limit(limit) => {
-                if let Some(fetch) = limit.fetch {
-                    if fetch == 0 {
+                // Only supports rewriting for literal fetch
+                let FetchType::Literal(fetch) = limit.get_fetch_type()? else {
+                    return Ok(Transformed::no(LogicalPlan::Limit(limit)));
+                };
+
+                if let Some(v) = fetch {
+                    if v == 0 {
                         return Ok(Transformed::yes(LogicalPlan::EmptyRelation(
                             EmptyRelation {
                                 produce_one_row: false,
                                 schema: Arc::clone(limit.input.schema()),
                             },
                         )));
                     }
-                } else if limit.skip == 0 {
-                    // input also can be Limit, so we should apply again.
+                } else if matches!(limit.get_skip_type()?, SkipType::Literal(0)) {
+                    // If fetch is `None` and skip is 0, then Limit takes no effect and
+                    // we can remove it. Its input also can be Limit, so we should apply again.
                     return Ok(self
                         .rewrite(Arc::unwrap_or_clone(limit.input), _config)
                         .unwrap());
```

Comment:
```suggestion
                        .rewrite(Arc::unwrap_or_clone(limit.input), _config)?);
```
Just noticed this point. It's better to return an error instead of a panic. Could you help to improve it?

Reply:
Done. Thank you @goldmedal